File: python/src/cryptopals/set2/challenge14.py
from random import randint

from Crypto.Cipher import AES
import secrets
from base64 import b64decode
from cryptopals.basics import padding, chunks

BLOCK_SIZE = 16

key = secrets.token_bytes(BLOCK_SIZE)
random_prefix = secrets.token_bytes(randint(1, 100))

target_bytes = b64decode(
    """
Um9sbGluJyBpbiBteSA1LjAKV2l0aCBteSByYWctdG9wIGRvd24gc28gbXkg
aGFpciBjYW4gYmxvdwpUaGUgZ2lybGllcyBvbiBzdGFuZGJ5IHdhdmluZyBq
dXN0IHRvIHNheSBoaQpEaWQgeW91IHN0b3A/IE5vLCBJIGp1c3QgZHJvdmUg
YnkK"""
)


def encryption_oracle(attacker_controlled):
    aes = AES.new(key, AES.MODE_ECB)
    return aes.encrypt(padding(random_prefix + attacker_controlled + target_bytes))


def lookup(skip, inject_length, known_secret: bytes, known_content: bytes):
    for b in range(256):
        secret = encryption_oracle((b"\x00" * inject_length) + known_content + bytes([b]))
        if secret[skip : skip + BLOCK_SIZE] == known_secret[skip : skip + BLOCK_SIZE]:
            return b


def find_ignorable_part(ignore_blocks):
    blank_blocks = b"\x00" * ignore_blocks * BLOCK_SIZE
    secret = encryption_oracle(blank_blocks)
    last_chunk = None
    for length_prefix, chunk in chunks(secret, BLOCK_SIZE):
        if last_chunk == chunk:
            # Magic Seven
            return length_prefix + 7 * BLOCK_SIZE
        last_chunk = chunk


def encryption_detector(printing=False):
    """
    >>> encryption_detector().startswith('Rollin')
    True
    """
    skip = find_ignorable_part(3)

    known = bytearray()
    for inject_length in range(159, -100, -1):
        secret = encryption_oracle(b"\x00" * inject_length)
        b = lookup(skip, inject_length, secret, known)
        if b is None:
            return known.decode()
        if b != 0:
            known.append(b)
            if printing:
                print(chr(b), end="")


if __name__ == "__main__":
    encryption_detector(True)

File: kotlin/src/main/kotlin/io/github/ocirne/cryptopals/set2/Challenge14.kt
package io.github.ocirne.cryptopals.set2

import io.github.ocirne.cryptopals.Basics
import io.github.ocirne.cryptopals.Basics.Extensions.repeat
import io.github.ocirne.cryptopals.Oracle
import io.github.ocirne.cryptopals.crypto.AES
import kotlin.random.Random

class Challenge14 {

    class PrefixSuffixEcbOracle(unknownString: String): Oracle {

        private val target = Basics.decodeFromBase64(unknownString)
        private val key = Random.nextBytes(AES.BLOCK_SIZE)
        private val aes = AES.ECB(key)
        private val lengthRandomPrefix = Random.nextInt(1,100)
        private val randomPrefix = Random.nextBytes(lengthRandomPrefix)

        override fun encrypt(attackerControlled: ByteArray): ByteArray {
            return aes.encrypt(randomPrefix + attackerControlled + target)
        }
    }

    class Attacker(val oracle: Oracle) {

        private fun lookup(skip: Int, injectLength: Int, knownSecret: ByteArray, knownContent: ByteArray): Byte? {
            return IntRange(0, 255)
                .map(Int::toByte)
                .firstOrNull { b ->
                    val secret = oracle.encrypt(0x00.toByte().repeat(injectLength) + knownContent + byteArrayOf(b))
                    secret.drop(skip).take(AES.BLOCK_SIZE) == knownSecret.drop(skip).take(AES.BLOCK_SIZE)
                }
        }

        private fun findIgnorablePart(): Int {
            val blankBlocks = 0x00.toByte().repeat(3 * AES.BLOCK_SIZE)
            val secret = oracle.encrypt(blankBlocks)
            val countIgnoreBlocks = secret.asIterable()
                .chunked(AES.BLOCK_SIZE)
                .zipWithNext()
                .indexOfFirst { (a, b) -> a == b }
            // TODO Magic eight
            return (countIgnoreBlocks + 8) * AES.BLOCK_SIZE
        }

        fun crack(): String {
            val skip = findIgnorablePart()
            var known = byteArrayOf()
            for (i in 159 downTo -100) {
                val secret = oracle.encrypt(0x00.toByte().repeat(i))
                val b = lookup(skip, i, secret, known) ?: return String(known)
                if (b != 0.toByte()) {
                    known += b
                }
            }
            return String(known)
        }
    }
}
